framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -272,19 +273,37 @@ private void refineArrayAtLengthAccess(FieldAccessNode arrayLengthNode, CFStore
         if (value == null) {
             return;
         }
+
+        boolean isIntRange = false;
+
         AnnotationMirror lengthAnno =
                 AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntVal.class);
 
+        if (lengthAnno == null) {
+            lengthAnno =
+                    AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntRange.class);
+            isIntRange = true;
+        }
+
         if (lengthAnno != null) {
-            List<Long> lengthValues = ValueAnnotatedTypeFactory.getIntValues(lengthAnno);
-            List<Integer> arrayLenValues = new ArrayList<>(lengthValues.size());
-            for (Long l : lengthValues) {
-                arrayLenValues.add(l.intValue());
+            Range range;
+            if (isIntRange) {
+                range = ValueAnnotatedTypeFactory.getRange(lengthAnno);
+            } else {
+                List<Long> lengthValues = ValueAnnotatedTypeFactory.getIntValues(lengthAnno);
+                range = new Range(Collections.min(lengthValues), Collections.max(lengthValues));
```

Comment:
This loses precision if the array.length is something like `@IntVal({1, 3})`.  Is there are reason for this?  You could use the RangeOrListOfValues class here.